File: labs/algorithms-live-coding/java/src/main/java/learning/livecoding/java/NumberOfIslandsExercise.java
package learning.livecoding.java;

public final class NumberOfIslandsExercise implements Exercise {
    @Override
    public String id() {
        return "number-of-islands";
    }

    @Override
    public String title() {
        return "Number of Islands";
    }

    @Override
    public String summary() {
        return "Grid DFS. Treat each unvisited land cell as the start of one component, then flood-fill it.";
    }

    int countIslands(char[][] grid) {
        int islands = 0;

        for (int row = 0; row < grid.length; row++) {
            for (int col = 0; col < grid[row].length; col++) {
                if (grid[row][col] == '1') {
                    islands++;
                    sink(grid, row, col);
                }
            }
        }

        return islands;
    }

    private void sink(char[][] grid, int row, int col) {
        if (row < 0 || row >= grid.length || col < 0 || col >= grid[row].length || grid[row][col] != '1') {
            return;
        }

        grid[row][col] = '0';
        sink(grid, row + 1, col);
        sink(grid, row - 1, col);
        sink(grid, row, col + 1);
        sink(grid, row, col - 1);
    }

    @Override
    public void run() {
        char[][] first = {
            {'1', '1', '0', '0', '0'},
            {'1', '1', '0', '0', '0'},
            {'0', '0', '1', '0', '0'},
            {'0', '0', '0', '1', '1'}
        };
        char[][] second = {
            {'1', '1', '1'},
            {'0', '1', '0'},
            {'1', '1', '1'}
        };

        ExerciseSupport.expectEquals("three-islands", 3, countIslands(first));
        ExerciseSupport.expectEquals("single-island", 1, countIslands(second));
    }
}

File: labs/algorithms-live-coding/kotlin/src/main/kotlin/learning/livecoding/kotlin/NumberOfIslandsExercise.kt
package learning.livecoding.kotlin

object NumberOfIslandsExercise : Exercise {
    override val id = "number-of-islands"
    override val title = "Number of Islands"
    override val summary = "Grid DFS. Treat each unvisited land cell as the start of one component, then flood-fill it."

    fun countIslands(grid: Array<CharArray>): Int {
        var islands = 0

        for (row in grid.indices) {
            for (col in grid[row].indices) {
                if (grid[row][col] == '1') {
                    islands++
                    sink(grid, row, col)
                }
            }
        }

        return islands
    }

    private fun sink(grid: Array<CharArray>, row: Int, col: Int) {
        if (row !in grid.indices || col !in grid[row].indices || grid[row][col] != '1') {
            return
        }

        grid[row][col] = '0'
        sink(grid, row + 1, col)
        sink(grid, row - 1, col)
        sink(grid, row, col + 1)
        sink(grid, row, col - 1)
    }

    override fun run() {
        val first = arrayOf(
            charArrayOf('1', '1', '0', '0', '0'),
            charArrayOf('1', '1', '0', '0', '0'),
            charArrayOf('0', '0', '1', '0', '0'),
            charArrayOf('0', '0', '0', '1', '1'),
        )
        val second = arrayOf(
            charArrayOf('1', '1', '1'),
            charArrayOf('0', '1', '0'),
            charArrayOf('1', '1', '1'),
        )

        ExerciseSupport.expectEquals("three-islands", 3, countIslands(first))
        ExerciseSupport.expectEquals("single-island", 1, countIslands(second))
    }
}
